### ISSUE TITLE
Crash on malformed zip archive

### ISSUE THREAD

**Issue by MarkusKrell** [weichsel/ZIPFoundation#179]
# Summary
When ZIPFoundation tries to uncompress a specific malformed zip file it crashes.

# Steps to Reproduce
Please use the following attached minimal zip file
[archive.zip](https://github.com/weichsel/ZIPFoundation/files/4860075/archive.zip). Try to decompress this file with ZIPFoundation. I simply used your provided sample at https://github.com/weichsel/ZIPFoundation#unzipping-archives 

# Expected Results
Error message is shown indicating that the file is malformed.

# Actual Results

```
Process 50 launched: '/swift/.build/debug/test' (x86_64)
Process 50 stopped
* thread #1, name = 'test', stop reason = signal SIGILL: illegal instruction operand
    frame #0: 0x000055555558956b test`Entry.CentralDirectoryStructure.init(data=46 bytes, provider=0x0000555555581680 test`partial apply forwarder for closure #1 (Swift.Int) throws -> Foundation.Data in static (extension in ZIPFoundation):Foundation.Data.readStruct<A where A: ZIPFoundation.DataSerializable>(from: Swift.UnsafeMutablePointer<__C._IO_FILE>, at: Swift.Int) -> Swift.Optional<A> at <compiler-generated>) at Entry.swift:324:84
   321          self.internalFileAttributes = data.scanValue(start: 36)
   322          self.externalFileAttributes = data.scanValue(start: 38)
   323          self.relativeOffsetOfLocalHeader = data.scanValue(start: 42)
-> 324          let additionalDataLength = Int(self.fileNameLength + self.extraFieldLength + self.fileCommentLength)
   325          guard let additionalData = try? provider(additionalDataLength) else { return nil }
   326          guard additionalData.count == additionalDataLength else { return nil }
   327          var subRangeStart = 0
Target 0: (test) stopped.
```

```
(lldb) frame variable
(Data) data = 46 bytes
() -> () provider = 0x0000555555581680 test`partial apply forwarder for closure #1 (Swift.Int) throws -> Foundation.Data in static (extension in ZIPFoundation):Foundation.Data.readStruct<A where A: ZIPFoundation.DataSerializable>(from: Swift.UnsafeMutablePointer<__C._IO_FILE>, at: Swift.Int) -> Swift.Optional<A> at <compiler-generated>
(ZIPFoundation.Entry.CentralDirectoryStructure) self = {
  centralDirectorySignature = 33639248
  versionMadeBy = 831
  versionNeededToExtract = 10
  generalPurposeBitFlag = 0
  compressionMethod = 0
  lastModFileTime = 43499
  lastModFileDate = 20705
  crc32 = 1396555892
  compressedSize = 5
  uncompressedSize = 5
  fileNameLength = 24840
  extraFieldLength = 24929
  fileCommentLength = 24929
  diskNumberStart = 24929
  internalFileAttributes = 24929
  externalFileAttributes = 2176057354
  relativeOffsetOfLocalHeader = 0
  fileNameData = 0 bytes
  extraFieldData = 0 bytes
  fileCommentData = 0 bytes
}
(Data) additionalData = 0 bytes
(Int) subRangeStart = 0
(Int) subRangeEnd = 0
(Int) additionalDataLength = <variable not available>
```

I think the crash could be due to the fact that 24929*3 is larger than UInt16.max.

# Regression & Version
Version was 0.9.11, I verified this on iOS 13.5.1 and Linux

# Related Link
https://github.com/weichsel/ZIPFoundation/blob/3f8516cef138413d60f33d8255eece869be1702b/Sources/ZIPFoundation/Entry.swift#L324

Please let me know in case I can provide more details.

Thank you for your effort.
Markus

**Comment by weichsel** [comment]
Hello Markus,

Thanks for providing the detailed error report.  
I was not able to reproduce the issue with the archive you attached when using the unzip code from the `README`. ZIP Foundation fails to initialize an `Archive` because it fails to find an `endOfCentralDirectoryRecord`.  

> I think the crash could be due to the fact that 24929*3 is larger than UInt16.max.

Good point. Are you using the `data` based initializer of `Archive`? Can you try to change the framework code so that the cast is performed on each `length` property individually:
```swift
let additionalDataLength = Int(self.fileNameLength) + Int(self.extraFieldLength) + Int(self.fileCommentLength)
```

**Comment by MarkusKrell** [comment]
Hello Thomas,

> Are you using the data based initializer of Archive?

Sorry not really sure what you mean in this case. I am using the attached zip file and the following basic structure:
```
cat Package.swift 
// swift-tools-version:5.0
import PackageDescription
let package = Package(
    name: "<Your Product Name>",
    dependencies: [
                .package(url: "https://github.com/weichsel/ZIPFoundation/", .upToNextMajor(from: "0.9.0"))
    ],
    targets: [
        .target(
                name: "test",
                dependencies: ["ZIPFoundation"]),
    ]
```

and 

```
# cat Sources/test/main.swift 
import ZIPFoundation
import Foundation
import Glibc

let fileManager = FileManager()
let currentWorkingPath = fileManager.currentDirectoryPath
var sourceURL = URL(fileURLWithPath: currentWorkingPath)
sourceURL.appendPathComponent("archive.zip")
var destinationURL = URL(fileURLWithPath: currentWorkingPath)
destinationURL.appendPathComponent("directory")
do {
    try fileManager.createDirectory(at: destinationURL, withIntermediateDirectories: true, attributes: nil)
    try fileManager.unzipItem(at: sourceURL, to: destinationURL)
} catch {
    print("Extraction of ZIP archive failed with error:\(error)")

```
> Can you try to change the framework code so that the cast is performed on each length property individually
Yes, I did it like you suggested and now what I get is:
```
root@ef962562d158:/swift# lldb .build/debug/test
(lldb) target create ".build/debug/test"
Current executable set to '/swift/.build/debug/test' (x86_64).
(lldb) run
Process 408 launched: '/swift/.build/debug/test' (x86_64)
Process 408 exited with status = 0 (0x00000000) 
```

So no crash anymore, however there is also no output that the file is corrupted. Of course, no output file is generated. Do I need to enable a debug flag for further details?

Kind regards,
Markus

**Comment by weichsel** [comment]
Markus,
I reviewed the code base for integer overflows that could happen for archives that contain entries with very long filename or extra data fields. There were some code paths that contained a similar bug to the one you pointed out (wrong add/cast order when adding/upcasting `UInt16` fields). I fixed those in PR #180 .

> So no crash anymore, however there is also no output that the file is corrupted. Of course, no output file is generated. Do I need to enable a debug flag for further details?

The archive initializer is failable and therefore returns `nil` when the archive is invalid (e.g. when it does not contain an `endOfCentralDirectory` record - like in the archive you provided).   

ZIP Foundation is strict when reading archives and does not try to skip or fix invalid entries.
